File: Reference/Lua-Game-AI-Programming/src/demo_framework/src/Agent.cpp
/**
 * Copyright (c) 2013 David Young dayoung@goliathdesigns.com
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 *  1. The origin of this software must not be misrepresented; you must not
 *  claim that you wrote the original software. If you use this software
 *  in a product, an acknowledgment in the product documentation would be
 *  appreciated but is not required.
 *
 *  2. Altered source versions must be plainly marked as such, and must not be
 *  misrepresented as being the original software.
 *
 *  3. This notice may not be removed or altered from any source
 *  distribution.
 */

#include "PrecompiledHeaders.h"

#include "demo_framework/include/Agent.h"
#include "demo_framework/include/AgentGroup.h"
#include "demo_framework/include/AgentUtilities.h"
#include "demo_framework/include/AnimationUtilities.h"
#include "demo_framework/include/LuaScriptUtilities.h"
#include "demo_framework/include/PhysicsUtilities.h"
#include "demo_framework/include/SandboxObject.h"
#include "demo_framework/include/SandboxUtilities.h"

namespace
{
inline Ogre::Quaternion BtQuaterionToQuaternion(const btQuaternion& quaternion)
{
    return Ogre::Quaternion(
        quaternion.w(), quaternion.x(), quaternion.y(), quaternion.z());
}

inline btQuaternion QuaternionToBtQuaternion(
    const Ogre::Quaternion& quaternion)
{
    return btQuaternion(quaternion.x, quaternion.y, quaternion.z, quaternion.w);
}

inline OpenSteer::Vec3 Vector3ToVec3(const Ogre::Vector3& vector)
{
    return OpenSteer::Vec3(
        static_cast<float>(vector.x),
        static_cast<float>(vector.y),
        static_cast<float>(vector.z));
}

inline Ogre::Vector3 Vec3ToVector3(const OpenSteer::Vec3& vector)
{
    return Ogre::Vector3(
        Ogre::Real(vector.x), Ogre::Real(vector.y), Ogre::Real(vector.z));
}

inline btVector3 Vector3ToBtVector3(const Ogre::Vector3& vector)
{
    return btVector3(
        btScalar(vector.x), btScalar(vector.y), btScalar(vector.z));
}
}  // anonymous namespace

// Some default humanistic values.
const float Agent::DEFAULT_AGENT_HEALTH =           100.0f;
const float Agent::DEFAULT_AGENT_HEIGHT =           1.6f;  // meters (5.2 feet)
const float Agent::DEFAULT_AGENT_MASS =             90.7f;  // kilograms (200 lbs)
const float Agent::DEFAULT_AGENT_MAX_FORCE =        1000.0f;  // newtons (kg*m/s^2)
const float Agent::DEFAULT_AGENT_MAX_SPEED =        7.0f;  // m/s (23.0 ft/s)
const float Agent::DEFAULT_AGENT_RADIUS =           0.3f;  // meters (1.97 feet)
const float Agent::DEFAULT_AGENT_SPEED =            0.0f;  // m/s (0 ft/s)
const float Agent::DEFAULT_AGENT_TARGET_RADIUS =    0.5f;  // meters (1.64 feet)
const float Agent::DEFAULT_AGENT_WALKABLE_CLIMB =   DEFAULT_AGENT_RADIUS / 2.0f;
const float Agent::DEFAULT_AGENT_WALKABLE_SLOPE =   45.0f;
const Ogre::String Agent::DEFAULT_AGENT_TEAM =      "team1";

Agent::Agent(
    const unsigned int agentId,
    Ogre::SceneNode* const sceneNode,
    btRigidBody* const rigidBody)
    : Object(agentId, Object::AGENT),
    sandbox_(NULL),
    health_(DEFAULT_AGENT_HEALTH),
    height_(DEFAULT_AGENT_HEIGHT),
    hasPath_(false),
    sceneNode_(sceneNode),
    maxForce_(DEFAULT_AGENT_MAX_FORCE),
    maxSpeed_(DEFAULT_AGENT_MAX_SPEED),
    mass_(DEFAULT_AGENT_MASS),
    radius_(DEFAULT_AGENT_RADIUS),
    speed_(DEFAULT_AGENT_SPEED),
    rigidBody_(rigidBody),
    targetRadius_(DEFAULT_AGENT_TARGET_RADIUS),
    team_(DEFAULT_AGENT_TEAM),
    target_(Ogre::Vector3::ZERO)
{
    assert(sceneNode_);
    // SceneNode shouldn't have attached entities already.
    assert(!sceneNode_->numAttachedObjects());
    // SceneNode shouldn't have attached children.
    assert(!sceneNode_->numChildren());

    luaVM_ = LuaScriptUtilities::CreateVM();

    // Add general C functions.
    LuaScriptUtilities::BindVMFunctions(luaVM_);

    // Add Agent specific functions.
    AgentUtilities::BindVMFunctions(luaVM_);

    // Add Sandbox specific functions.
    SandboxUtilities::BindVMFunctions(luaVM_);

    // Add Animation specific functions.
    AnimationUtilities::BindVMFunctions(luaVM_);

    if (rigidBody_)
    {
        PhysicsUtilities::SetRigidBodyMass(rigidBody_, DEFAULT_AGENT_MASS);
        rigidBody->setUserPointer(this);
    }

    SetForward(Ogre::Vector3::UNIT_Z);
}

Agent::~Agent()
{
    sandbox_ = NULL;
    sceneNode_ = NULL;

    if (rigidBody_)
    {
        PhysicsUtilities::DeleteRigidBody(rigidBody_);
    }

    rigidBody_ = NULL;

    LuaScriptUtilities::DestroyVM(luaVM_);
}

void Agent::Cleanup()
{
    AgentUtilities::CallLuaAgentCleanup(this);
}

void Agent::ClearPath()
{
    hasPath_ = false;
}

Ogre::Vector3 Agent::ForceToAlign(
    const Ogre::Real maxDistance,
    const Ogre::Degree maxAngle,
    const AgentGroup& group)
{
    const float maxCosAngle = cosf(maxAngle.valueRadians());
    return Vec3ToVector3(steerForAlignment(maxDistance, maxCosAngle, group));
}

Ogre::Vector3 Agent::ForceToAvoidAgents(
    const std::vector<Agent*>& agents, const float predictionTime)
{
    const static float MIN_PREDICTION_TIME = 0.1f;

    OpenSteer::AVGroup group;
    std::vector<Agent*>::const_iterator it;

    // TODO(David Young): Avoid copying into the AVGroup.
    for (it = agents.begin(); it != agents.end(); ++it)
    {
        Agent* const agent = *it;
        group.push_back(agent);
    }

    return Vec3ToVector3(steerToAvoidNeighbors(
        std::max(MIN_PREDICTION_TIME, predictionTime), group));
}

Ogre::Vector3 Agent::ForceToAvoidObjects(
    const std::map<unsigned int, SandboxObject*>& objects,
    const float predictionTime)
{
    const static float MIN_PREDICTION_TIME = 0.1f;
    const float timeToCollision = std::max(MIN_PREDICTION_TIME, predictionTime);

    OpenSteer::ObstacleGroup group;
    std::map<unsigned int, SandboxObject*>::const_iterator it;

    OpenSteer::Vec3 avoidForce = OpenSteer::Vec3::zero;

    // TODO(David Young): Avoid copying into the ObstacleGroup.
    for (it = objects.begin(); it != objects.end(); ++it)
    {
        SandboxObject* const object = it->second;

        // Only avoid objects that aren't fixed.
        if (object->GetMass() > 0)
        {
            avoidForce += object->steerToAvoid(*this, timeToCollision);
        }
    }

    return Vec3ToVector3(avoidForce);
}

Ogre::Vector3 Agent::ForceToCombine(
    const Ogre::Real maxDistance,
    const Ogre::Degree maxAngle,
    const AgentGroup& group)
{
    const float maxCosAngle = cosf(maxAngle.valueRadians());
    return Vec3ToVector3(steerForCohesion(maxDistance, maxCosAngle, group));
}

Ogre::Vector3 Agent::ForceToFleePosition(const Ogre::Vector3& position)
{
    return Vec3ToVector3(steerForFlee(Vector3ToVec3(position)));
}

Ogre::Vector3 Agent::ForceToFollowPath(const float predicitionTime)
{
    return ForceToFollowPath(path_, predicitionTime);
}

Ogre::Vector3 Agent::ForceToFollowPath(
    AgentPath& path, const float predictionTime)
{
    const static int FORWARD_DIRECTION = 1;
    const static float MIN_PREDICTION_TIME = 0.1f;

    if (path.GetNumberOfPathPoints())
    {
        return Vec3ToVector3(steerToFollowPath(
            FORWARD_DIRECTION,
            std::max(MIN_PREDICTION_TIME, predictionTime),
            path));
    }
    return Ogre::Vector3(0.0f);
}

Ogre::Vector3 Agent::ForceToPosition(const Ogre::Vector3& position)
{
    return Vec3ToVector3(steerForSeek(Vector3ToVec3(position)));
}

Ogre::Vector3 Agent::ForceToStayOnPath(const float predictionTime)
{
    return ForceToStayOnPath(path_, predictionTime);
}

Ogre::Vector3 Agent::ForceToSeparate(
    const Ogre::Real maxDistance,
    const Ogre::Degree maxAngle,
    const AgentGroup& group)
{
    const float maxCosAngle = cosf(maxAngle.valueRadians());
    return Vec3ToVector3(steerForSeparation(maxDistance, maxCosAngle, group));
}

Ogre::Vector3 Agent::ForceToStayOnPath(
    AgentPath& path, const float predictionTime)
{
    const static float MIN_PREDICTION_TIME = 0.1f;

    if (path.GetNumberOfPathPoints())
    {
        return Vec3ToVector3(steerToStayOnPath(
        std::max(MIN_PREDICTION_TIME, predictionTime), path));
    }

    return Ogre::Vector3(0.0f);
}

Ogre::Vector3 Agent::ForceToTargetSpeed(const Ogre::Real speed)
{
    return Vec3ToVector3(steerForTargetSpeed(float(speed)));
}

Ogre::Vector3 Agent::ForceToWander(const Ogre::Real deltaMilliseconds)
{
    return Vec3ToVector3(steerForWander(deltaMilliseconds));
}

OpenSteer::Vec3 Agent::forward() const
{
    return Vector3ToVec3(GetForward());
}

Ogre::Vector3 Agent::GetForward() const
{
    if (rigidBody_)
    {
        const btQuaternion quaterion = rigidBody_->getOrientation();
        return Ogre::Quaternion(
            quaterion.w(), quaterion.x(), quaterion.y(), quaterion.z()).zAxis();
    }
    else if (sceneNode_)
    {
        return sceneNode_->getOrientation().zAxis();
    }

    return Ogre::Vector3::UNIT_Z;
}

Ogre::Real Agent::GetHealth() const
{
    return health_;
}

Ogre::Real Agent::GetHeight() const
{
    return height_;
}

Ogre::Vector3 Agent::GetLeft() const
{
    if (rigidBody_)
    {
        const btQuaternion quaterion = rigidBody_->getOrientation();
        return Ogre::Quaternion(
            quaterion.w(), quaterion.x(), quaterion.y(), quaterion.z()).xAxis();
    }
    else if (sceneNode_)
    {
        return sceneNode_->getOrientation().xAxis();
    }

    return Ogre::Vector3::UNIT_X;
}

lua_State* Agent::GetLuaVM()
{
    return luaVM_;
}

Ogre::Real Agent::GetMass() const
{
    if (rigidBody_)
    {
        return Ogre::Real(PhysicsUtilities::GetRigidBodyMass(rigidBody_));
    }
    else
    {
        return mass_;
    }
}

Ogre::Real Agent::GetMaxForce() const
{
    return maxForce_;
}

Ogre::Real Agent::GetMaxSpeed() const
{
    return maxSpeed_;
}

Ogre::Quaternion Agent::GetOrientation() const
{
    if (rigidBody_)
    {
        return BtQuaterionToQuaternion(rigidBody_->getOrientation());
    }
    else if (sceneNode_)
    {
        return sceneNode_->_getDerivedOrientation();
    }

    return Ogre::Quaternion::ZERO;
}

AgentPath Agent::GetPath()
{
    return path_;
}

const AgentPath& Agent::GetPath() const
{
    return path_;
}

Ogre::Vector3 Agent::GetPosition() const
{
    if (rigidBody_)
    {
        return PhysicsUtilities::BtVector3ToVector3(
            rigidBody_->getCenterOfMassPosition());
    }
    else if (sceneNode_)
    {
        return sceneNode_->_getDerivedPosition();
    }

    return Ogre::Vector3::ZERO;
}

Ogre::Real Agent::GetRadius() const
{
    return radius_;
}

btRigidBody* Agent::GetRigidBody()
{
    return rigidBody_;
}

Sandbox* Agent::GetSandbox()
{
    return sandbox_;
}

Ogre::SceneNode* Agent::GetSceneNode()
{
    return sceneNode_;
}

Ogre::Real Agent::GetSpeed() const
{
    if (rigidBody_)
    {
        const btVector3 velocity = rigidBody_->getLinearVelocity();
        // Ignore movement along the y axis.
        // Gravity does not contribute to the speed of the Agent.
        return Ogre::Real(
            Ogre::Vector3(velocity.x(), 0, velocity.z()).length());
    }
    else
    {
        return speed_;
    }
}

Ogre::Vector3 Agent::GetTarget() const
{
    return target_;
}

Ogre::Real Agent::GetTargetRadius() const
{
    return targetRadius_;
}

const Ogre::String& Agent::GetTeam() const
{
    return team_;
}

Ogre::Vector3 Agent::GetUp() const
{
    if (rigidBody_)
    {
        const btQuaternion quaterion = rigidBody_->getOrientation();
        return Ogre::Quaternion(
            quaterion.w(), quaterion.x(), quaterion.y(), quaterion.z()).yAxis();
    }
    else if (sceneNode_)
    {
        return sceneNode_->getOrientation().yAxis();
    }

    return Ogre::Vector3::UNIT_Y;
}

Ogre::Vector3 Agent::GetVelocity() const
{
    if (rigidBody_)
    {
        const btVector3 velocity = rigidBody_->getLinearVelocity();

        return Ogre::Vector3(velocity.x(), velocity.y(), velocity.z());
    }
    return GetForward() * speed_;
}

OpenSteer::Vec3 Agent::globalizeDirection(
    const OpenSteer::Vec3& localDirection) const
{
    (void)localDirection;
    // not implemented
    assert(false);

    return OpenSteer::Vec3();
}
OpenSteer::Vec3 Agent::globalizePosition(
    const OpenSteer::Vec3& localPosition) const
{
    (void)localPosition;
    // not implemented
    assert(false);

    return OpenSteer::Vec3();
}

OpenSteer::Vec3 Agent::globalRotateForwardToSide(
    const OpenSteer::Vec3& globalForward) const
{
    (void)globalForward;
    // not implemented
    assert(false);

    return OpenSteer::Vec3();
}

bool Agent::HasPath() const
{
    return hasPath_;
}

void Agent::Initialize()
{
    AgentUtilities::CreateRigidBodyCapsule(this);

    AgentUtilities::CallLuaAgentInitialize(this);
}

void Agent::LoadScript(
    const char* const luaScript,
    const size_t bufferSize,
    const char* const fileName)
{
    AgentUtilities::LoadScript(this, luaScript, bufferSize, fileName);
}

OpenSteer::Vec3 Agent::localizeDirection(
    const OpenSteer::Vec3& globalDirection) const
{
    // TODO(David Young): This is very slow, convert to native Ogre math.
    return OpenSteer::Vec3(
        globalDirection.dot(Vector3ToVec3(GetLeft())),
        globalDirection.dot(Vector3ToVec3(GetUp())),
        globalDirection.dot(Vector3ToVec3(GetForward())));
}

OpenSteer::Vec3 Agent::localizePosition(
    const OpenSteer::Vec3& globalPosition) const
{
    OpenSteer::Vec3 globalOffset =
        globalPosition - Vector3ToVec3(GetPosition());

    return localizeDirection (globalOffset);
}

OpenSteer::Vec3 Agent::localRotateForwardToSide(
    const OpenSteer::Vec3& side) const
{
    (void)side;
    // not implemented
    assert(false);

    return OpenSteer::Vec3();
}

float Agent::mass() const
{
    return static_cast<float>(GetMass());
}

float Agent::maxForce() const
{
    return static_cast<float>(GetMaxForce());
}

float Agent::maxSpeed() const
{
    return static_cast<float>(GetMaxSpeed());
}

OpenSteer::Vec3 Agent::position() const
{
    return Vector3ToVec3(GetPosition());
}

OpenSteer::Vec3 Agent::predictFuturePosition(
    const float predictionTime) const
{
    return Vector3ToVec3(PredictFuturePosition(Ogre::Real(predictionTime)));
}

Ogre::Vector3 Agent::PredictFuturePosition(
    const Ogre::Real predictionTime) const
{
    return GetPosition() + GetVelocity() * std::max(Ogre::Real(0), predictionTime);
}

float Agent::radius() const
{
    return static_cast<float>(GetRadius());
}

void Agent::regenerateOrthonormalBasisUF(
    const OpenSteer::Vec3& newUnitForward)
{
    (void)newUnitForward;
    // not implemented
    assert(false);
}

void Agent::regenerateOrthonormalBasis(const OpenSteer::Vec3& newForward)
{
    (void)newForward;
    // not implemented
    assert(false);
}

void Agent::regenerateOrthonormalBasis(
    const OpenSteer::Vec3& newForward, const OpenSteer::Vec3& newUp)
{
    (void)newForward;
    (void)newUp;
    // not implemented
    assert(false);
}

bool Agent::ReloadScript(
    const char* const luaScript,
    const size_t bufferSize,
    const char* const fileName)
{
    (void)luaScript;
    (void)bufferSize;
    (void)fileName;

    return false;
}

void Agent::RemovePath()
{
    hasPath_ = false;
}

void Agent::resetLocalSpace()
{
    // not implemented
    assert(false);
}

bool Agent::rightHanded() const
{
    return true;
}

void Agent::SetForward(const Ogre::Quaternion& orientation)
{
    if (rigidBody_)
    {
        PhysicsUtilities::SetRigidBodyOrientation(
            rigidBody_, QuaternionToBtQuaternion(orientation));
    }

    if (sceneNode_)
    {
        sceneNode_->setOrientation(orientation);
    }
}

void Agent::SetForward(const Ogre::Vector3& forward)
{
    Ogre::Vector3 up = Ogre::Vector3::UNIT_Y;

    const Ogre::Vector3 zAxis = forward.normalisedCopy();
    const Ogre::Vector3 xAxis = up.crossProduct(zAxis);
    const Ogre::Vector3 yAxis = zAxis.crossProduct(xAxis);

    Ogre::Quaternion orientation(xAxis, yAxis, zAxis);

    // Update both the rigid body and scene node.
    if (rigidBody_)
    {
        PhysicsUtilities::SetRigidBodyOrientation(
            rigidBody_, QuaternionToBtQuaternion(orientation));
    }

    if (sceneNode_)
    {
        sceneNode_->setOrientation(orientation);
    }
}

OpenSteer::Vec3 Agent::setForward(OpenSteer::Vec3 forward)
{
    (void)forward;
    // not implemented
    assert(false);

    return OpenSteer::Vec3();
}

void Agent::SetHealth(const Ogre::Real health)
{
    health_ = health;
}

void Agent::SetHeight(const Ogre::Real height)
{
    height_ = std::max(Ogre::Real(0), height);

    if (rigidBody_)
    {
        AgentUtilities::UpdateRigidBodyCapsule(this);
    }
}

float Agent::setMass(float mass)
{
    SetMass(Ogre::Real(mass));
    return this->mass();
}

void Agent::SetMass(const Ogre::Real mass)
{
    if (rigidBody_)
    {
        PhysicsUtilities::SetRigidBodyMass(
            rigidBody_, std::max(Ogre::Real(0), mass));
    }

    mass_ = mass;
}

float Agent::setMaxForce(float force)
{
    SetMaxForce(Ogre::Real(force));
    return maxForce();
}

void Agent::SetMaxForce(const Ogre::Real force)
{
    maxForce_ = std::max(Ogre::Real(0), force);
}

float Agent::setMaxSpeed(float speed)
{
    SetMaxSpeed(Ogre::Real(speed));
    return maxSpeed();
}

void Agent::SetMaxSpeed(const Ogre::Real speed)
{
    maxSpeed_ = std::max(Ogre::Real(0), speed);
}

void Agent::SetPath(const AgentPath& path)
{
    path_ = path;
    hasPath_ = true;
}

OpenSteer::Vec3 Agent::setPosition(OpenSteer::Vec3 position)
{
    SetPosition(Vec3ToVector3(position));
    return this->position();
}

void Agent::SetPosition(const Ogre::Vector3 position)
{
    if (rigidBody_)
    {
        PhysicsUtilities::SetRigidBodyPosition(
            rigidBody_, Vector3ToBtVector3(position));
    }

    if (sceneNode_)
    {
        sceneNode_->setPosition(position);
    }
}

float Agent::setRadius(float radius)
{
    SetRadius(Ogre::Real(radius));
    return this->radius();
}

void Agent::SetRadius(const Ogre::Real radius)
{
    if (sceneNode_)
    {
        radius_ = std::max(Ogre::Real(0), radius);
    }

    if (rigidBody_)
    {
        AgentUtilities::UpdateRigidBodyCapsule(this);
    }
}

void Agent::SetRigidBody(btRigidBody* const rigidBody)
{
    rigidBody_ = rigidBody;

    if (rigidBody_)
    {
        rigidBody_->setUserPointer(this);
    }
}

void Agent::SetSandbox(Sandbox* const sandbox)
{
    sandbox_ = sandbox;
}

OpenSteer::Vec3 Agent::setSide(OpenSteer::Vec3 side)
{
    (void)side;
    // not implemented
    assert(false);

    return OpenSteer::Vec3();
}

float Agent::setSpeed(float speed)
{
    SetSpeed(Ogre::Real(speed));
    return this->speed();
}

void Agent::SetSpeed(const Ogre::Real speed)
{
    // TODO(David Young): Need to update rigidbody's velocity based on input.
    speed_ = speed;
}

void Agent::SetTarget(const Ogre::Vector3& target)
{
    target_ = target;
}

void Agent::SetTargetRadius(const Ogre::Real radius)
{
    targetRadius_ = std::max(Ogre::Real(0), radius);
}

void Agent::SetTeam(const Ogre::String& team)
{
    team_ = team;
}

void Agent::setUnitSideFromForwardAndUp()
{
    // not implemented
    assert(false);
}

OpenSteer::Vec3 Agent::setUp(OpenSteer::Vec3 up)
{
    (void)up;
    // not implemented
    assert(false);

    return OpenSteer::Vec3();
}

void Agent::SetVelocity(const Ogre::Vector3& velocity)
{
    if (rigidBody_)
    {
        PhysicsUtilities::SetRigidBodyVelocity(
            rigidBody_, Vector3ToBtVector3(velocity));
    }

    SetSpeed(Ogre::Vector3(velocity.x, 0, velocity.z).length());
}

OpenSteer::Vec3 Agent::side() const
{
    return Vector3ToVec3(GetLeft());
}

float Agent::speed() const
{
    return static_cast<float>(GetSpeed());
}

OpenSteer::Vec3 Agent::up() const
{
    return Vector3ToVec3(GetUp());
}

void Agent::update(const float currentTime, const float elapsedTime)
{
    (void)currentTime;
    (void)elapsedTime;
    // not implemented
    assert(false);
}

void Agent::Update(const int deltaMilliseconds)
{
    AgentUtilities::CallLuaAgentUpdate(this, deltaMilliseconds);
    AgentUtilities::UpdateWorldTransform(this);

    if (rigidBody_)
    {
        const btVector3 velocity = rigidBody_->getLinearVelocity();
        SetSpeed(Ogre::Vector3(velocity.x(), 0, velocity.z()).length());
    }
}

OpenSteer::Vec3 Agent::velocity() const
{
    return Vector3ToVec3(GetVelocity());
}

File: Reference/Lua-Game-AI-Programming/script/Agent.lua
-- Agent.lua
Agent = {}
--- @summary Metamethod that determines if the agent is equal to the other
---   variable.
--- @param agent Agent to compare.
--- @param variable Variable to compare against.
--- @return boolean True if the variable is equal to the agent.
--- @package Agent
--- @example comparison = agent == variable;
function Agent.__eq(agent)
end
function Agent.__index(agent)
end
function Agent.__towatch(agent)
end
--- @summary Apply a three dimensional force in meters to the agent.
--- @param agent Agent to apply force on.
--- @param vector Representing force in meters.
--- @package Agent
--- @example force = Agent.ApplyForce(agent, Vector.new(1, 0, 0));
function Agent.ApplyForce(agent)
end
--- @summary Calculate a force vector to steer the agent into aligning with
---   neighbor agents.
--- @param agent Agent to calculate a steering force for.
--- @param number Maximum distance to maintain with neighbors in meters.
--- @param number Maximum angle to maintain with neighbors in degrees.
--- @param table Table of agents indexed by number 1..n
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToAlign(agent, 1, 90, { agent2, agent3 });
function Agent.ForceToAlign(agent)
end
--- @summary Calculate a force to avoid all other agents within the sandbox.
--- @param agent Agent to calculate avoidance force for.
--- @param number Optional time in seconds to predict future movements, defaults
---   to 0.1 seconds.
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToAvoidAgents(agent);
--- @example force = Agent.ForceToAvoidAgents(agent, 1);
function Agent.ForceToAvoidAgents(agent)
end
--- @summary Calculate a force to avoid all movable objects within the sandbox.
--- @param agent Agent to calculate avoidance force for.
--- @param number Optional time in seconds to predict future movements, defaults
---   to 0.1 seconds.
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToAvoidObjects(agent);
--- @example force = Agent.ForceToAvoidObjects(agent, 1);
function Agent.ForceToAvoidObjects(agent)
end
--- @summary Calculate a force vector to steer the agent into moving toward a
---   group of agents.
--- @param agent Agent to calculate a steering force for.
--- @param number Agents must be within this distance to be considered within the
---   group.
--- @param number Agents must be within this degree of difference to be
---   considered within the group.
--- @param table Table of agents indexed by number 1..n
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToCombine(agent, 1, 90, { agent2, agent3 });
function Agent.ForceToCombine(agent)
end
--- @summary Calculate a force vector to steer the agent away from the position.
--- @param agent Agent to calculate a steering force for.
--- @param vector Position to steering away from.
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToFleePosition(agent, vector.new(10, 0, 10));
function Agent.ForceToFleePosition(agent)
end
--- @summary Calculate a force vector to steer toward the Agent's current path.
---   If the Agent doesn't have a path, returns the zero vector.
--- @param agent Agent to calculate following force for.
--- @param number Optional time in seconds to predict future movements, defaults
---   to 0.1 seconds.
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToFollowPath(agent);
--- @example force = Agent.ForceToFollowPath(agent, 1);
function Agent.ForceToFollowPath(agent)
end
--- @summary Calculate a force vector to steer the Agent's toward a position.
--- @param agent Agent to calculate seeking force for.
--- @param vector Position to steer towards.
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToPosition(agent, Vector.new(10, 0, 10));
function Agent.ForceToPosition(agent)
end
--- @summary Calculate a force vector to steer away from a group of Agents.
--- @param agent Agent to calculate a separation force for.
--- @param number Agents must be within this distance to be considered within the
---   group.
--- @param number Agents must be within this degree of difference to be
---   considered within the group.
--- @param table Table of agents indexed by number 1..n
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToSeparate(agent, 1, 90, { agent2, agent3 });
function Agent.ForceToSeparate(agent)
end
--- @summary Calculate a force vector to steer toward the Agent toward the
---   nearest path segment of their path.
--- @param agent Agent to calculate a separation force for.
--- @param number Optional time in seconds to predict future movements, defaults
---   to 0.1 seconds.
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToStayOnPath(agent);
--- @example force = Agent.ForceToStayOnPath(agent, 1);
function Agent.ForceToStayOnPath(agent)
end
--- @summary Calculate a force vector to accelerate or decelerate the Agent to
---   the specified speed.
--- @param agent Agent to calculate a speed adjustment force for.
--- @param number Speed in meters for the Agent to match.
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToTargetSpeed(agent, 3);
function Agent.ForceToTargetSpeed(agent)
end
--- @summary Calculate a force vector that randomly moves the agent in any
---   direction.  The delta time in milliseconds controls the magnitude of the
---   force vector.
--- @param agent Agent to calculate a force vector for.
--- @param number Delta time in milliseconds since the previous calling frame.
--- @return vector Force vector in meters.
--- @package Agent
--- @example force = Agent.ForceToWander(agent, deltaTimeInMillis);
function Agent.ForceToWander(agent)
end
function Agent.GetDistanceAlongPath(agent)
end
--- @summary Returns the forward axis of the Agent.
--- @param agent Agent to return the forward axis to.
--- @return vector Normalized forward vector.
--- @package Agent
--- @example forward = Agent.GetForward(agent);
function Agent.GetForward(agent)
end
--- @summary Returns the current health of the Agent.  Defaults to 100.
--- @param agent Agent to return the health of.
--- @return number Health of the Agent.
--- @package Agent
--- @example health = Agent.GetHealth(agent);
function Agent.GetHealth(agent)
end
--- @summary Returns the current height of the Agent.  Defaults to 1.6 meters.
--- @param agent Agent to return the height of.
--- @return number Height in meters.
--- @package Agent
--- @example height = Agent.GetHeight(agent);
function Agent.GetHeight(agent)
end
--- @summary Returns the unique id of the Agent.
--- @param agent Agent to return the id of.
--- @return number Unique id number.
--- @package Agent
--- @example id = Agent.GetId(agent);
function Agent.GetId(agent)
end
--- @summary Returns the left axis of the Agent.
--- @param agent Agent to return the left axis to.
--- @return vector Normalized left vector.
--- @package Agent
--- @example left = Agent.GetLeft(agent);
function Agent.GetLeft(agent)
end
--- @summary Returns the current mass of the Agent.  Defaults to 90.7 kilograms.
--- @param agent Agent to return the mass of.
--- @return number Mass of the Agent.
--- @package Agent
--- @example mass = Agent.GetMass(agent);
function Agent.GetMass(agent)
end
--- @summary Returns the maximum number of newtons the Agent's force vector can
---   reach.
--- @param agent Agent to return the maximum force of.
--- @return number Maximum force in number of newtons.
--- @package Agent
--- @example force = Agent.GetForce(agent);
function Agent.GetMaxForce(agent)
end
--- @summary Returns the maximum speed in meters per second the Agent can reach.
--- @param agent Agent to return the maximum speed of.
--- @return number Speed in meters per seconds.
--- @package Agent
--- @example speed = Agent.GetMaxSpeed(agent);
function Agent.GetMaxSpeed(agent)
end
function Agent.GetNearestPointOnPath(agent)
end
--- @summary Returns a table of vectors representing the current path of the
---   Agent.
--- @param agent Agent to return the path of.
--- @return table Table of vectors.  An empty table is returned if the Agent
---   has no path.
--- @package Agent
--- @example path = Agent.GetPath(agent);
function Agent.GetPath(agent)
end
function Agent.GetPointOnPath(agent)
end
--- @summary Returns the Agent's current position.  This is the Agent's
---   midpoint.
--- @param agent Agent to return the position of.
--- @return vector Position in meters.
--- @package Agent
--- @example position = Agent.GetPosition(agent);
function Agent.GetPosition(agent)
end
--- @summary Returns the Agent's current radius.  This is the Agent's capsule
---   radius used for avoidance and physics.
--- @param agent Agent to return the radius of.
--- @return number Radius in meters.
--- @package Agent
--- @example radius = Agent.GetRadius(agent);
function Agent.GetRadius(agent)
end
--- @summary Returns the Sandbox instance the Agent belongs to.
--- @param agent Agent to return the Sandbox from.
--- @return sandbox Sandbox instance.
--- @package Agent
--- @example sandbox = Agent.GetSandbox(agent);
function Agent.GetSandbox(agent)
end
--- @summary Returns the current speed of the agent as a scalar in meters per
---   second.
--- @param agent Agent to return the current speed of.
--- @return number Speed in meters per second.
--- @package Agent
--- @example speed = Agent.GetSpeed(agent);
function Agent.GetSpeed(agent)
end
--- @summary Returns the current target position of the agent in meters.
--- @param agent Agent to return the target of.
--- @return vector Position in meters.
--- @package Agent
--- @example target = Agent.GetTarget(agent);
function Agent.GetTarget(agent)
end
--- @summary Returns the current target radius of the agent in meters.  The
---   target radius is used to determine if the Agent is at the Agent's target
---   position.
--- @param agent Agent to return the target radius of.
--- @return number Target radius in meters.
--- @package Agent
--- @example targetRadius = Agent.GetTargetRadius(agent);
function Agent.GetTargetRadius(agent)
end
--- @summary Returns the current team name of the Agent, defaults to the empty
---   string.
--- @param agent Agent to return the team for.
--- @return string Team name.
--- @package Agent
--- @example team = Agent.GetTeam(agent);
function Agent.GetTeam(agent)
end
--- @summary Return the normalized up vector for the Agent.
--- @param agent Agent to return the up vector for.
--- @return vector Vector pointing in the up direction, normalized.
--- @package Agent
--- @example up = Agent.GetUp(agent);
function Agent.GetUp(agent)
end
--- @summary Returns the current velocity vector of the Agent.  The velocity
---   vector is the direction the Agent is moving and the magnitude of the vector
---   is the current speed of the Agent.  Velocity is measured in meters per
---   second.
--- @param agent Agent to return the velocity for.
--- @return vector Velocity vector in meters per second.
--- @package Agent
--- @example velocity = Agent.GetVelocity(agent);
function Agent.GetVelocity(agent)
end
function Agent.HasPath(agent)
end
function Agent.IsAgent(agent)
end
function Agent.PredictFuturePosition(agent)
end
function Agent.RemovePath(agent)
end
function Agent.RemovePhysics(agent)
end
function Agent.SetForward(agent)
end
function Agent.SetHealth(agent)
end
function Agent.SetHeight(agent)
end
function Agent.SetMass(agent)
end
function Agent.SetMaxForce(agent)
end
function Agent.SetMaxSpeed(agent)
end
function Agent.SetPath(agent)
end
function Agent.SetPosition(agent)
end
function Agent.SetRadius(agent)
end
function Agent.SetSpeed(agent)
end
function Agent.SetTarget(agent)
end
function Agent.SetTargetRadius(agent)
end
function Agent.SetTeam(agent)
end
function Agent.SetVelocity(agent)
end
return Agent